src/note_encryption.rs

```diff
@@ -312,26 +357,73 @@ impl ShieldedOutput<OrchardDomain, COMPACT_NOTE_SIZE> for CompactAction {
 
 #[cfg(test)]
 mod tests {
+    use proptest::prelude::*;
     use rand::rngs::OsRng;
     use zcash_note_encryption::{
-        try_compact_note_decryption, try_note_decryption, try_output_recovery_with_ovk,
+        try_compact_note_decryption, try_note_decryption, try_output_recovery_with_ovk, Domain,
         EphemeralKeyBytes,
     };
 
-    use super::{prf_ock_orchard, CompactAction, OrchardDomain, OrchardNoteEncryption};
+    use crate::note::note_type::testing::arb_note_type;
     use crate::note::NoteType;
     use crate::{
         action::Action,
         keys::{
             DiversifiedTransmissionKey, Diversifier, EphemeralSecretKey, IncomingViewingKey,
             OutgoingViewingKey,
         },
-        note::{ExtractedNoteCommitment, Nullifier, RandomSeed, TransmittedNoteCiphertext},
+        note::{
+            testing::arb_note, ExtractedNoteCommitment, Nullifier, RandomSeed,
+            TransmittedNoteCiphertext,
+        },
         primitives::redpallas,
         value::{NoteValue, ValueCommitment},
         Address, Note,
     };
 
+    use super::{
+        orchard_parse_note_plaintext_without_memo, prf_ock_orchard, CompactAction, OrchardDomain,
+        OrchardNoteEncryption,
+    };
+
+    proptest! {
+    #[test]
+    fn test_encoding_roundtrip(
+        note in arb_note(NoteValue::from_raw(10)),
+        note_type in arb_note_type(),
+    ) {
+        let memo = &crate::test_vectors::note_encryption::test_vectors()[0].memo;
+
+        // Encode.
+        let plaintext = OrchardDomain::note_plaintext_bytes(&note, &note.recipient(), memo);
+
+        // Decode.
+        let domain = OrchardDomain { rho: note.rho() };
+        let parsed_version = plaintext.0[0];
+        let parsed_memo = domain.extract_memo(&plaintext);
+
+        let (parsed_note, parsed_recipient) = orchard_parse_note_plaintext_without_memo(&domain, &plaintext.0,
+            |diversifier| {
+                assert_eq!(diversifier, &note.recipient().diversifier());
+                Some(*note.recipient().pk_d())
+            }
+        ).expect("Plaintext parsing failed");
+
+        // Check.
+        assert_eq!(parsed_note, note);
+        assert_eq!(parsed_recipient, note.recipient());
+        if note_type.is_native().into() {
```

Comment:
should be `parsed_note.note_type.is_native().into()`.

Reply:
Fixed using `prop::bool::ANY`